src/posts/comparing-methods-for-accessing-secrets-in-vault-from-kubernetes.md

```diff
@@ -0,0 +1,287 @@
+
+#### Vault REST API: Support
+
+When using the REST API, you can do anything that Vault, your programming language, and runtime support. However, there is not built-in support for synchronising secrets to Kubernetes secrets, nor any kind of secrets rotation.
+
+|  | Secrets Engines | Auth Methods | Sync to K8s Secrets | Secret Rotation |
+| --- | --- | --- | --- | --- |
+| Vault REST API | All | Any | No | No, not built-in |
+
+### Vault Sidecar Injector
+
+The [Vault Agent Sidecar Injector](https://developer.hashicorp.com/vault/docs/platform/k8s/injector) relies on [pod annotations](https://developer.hashicorp.com/vault/docs/platform/k8s/injector#secrets-via-annotations) and "injects" secrets into the pod. Under the hood there is a [Kubernetes Mutating Webhook Controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) that modifies the pod if the `vault.hashicorp.com/agent-inject: true` annotation is detected. A shared memory volume is mounted and used by sidecars (`init` and `sidecar`) that fetch and can also renew the secret.
+
+![k8s-vault-sidecar-injector](/blogs/comparing-methods-for-accessing-secrets-in-vault-from-kubernetes/k8s-vault-sidecar-injector.png)
+
+#### Vault Sidecar Injector: Effort
+
+**Platform Effort**: The Vault Sidecar Controller must be installed on each cluster that wants to use this method.
+
+**Deployment Effort**: Writing annotations on the pod will be the bulk of the work with this approach, and the annotations can be a little bit clumsy (especially as [there are so many](https://developer.hashicorp.com/vault/docs/platform/k8s/injector/annotations)) and can take a few tries to get it right.
+
+**Application Effort**: The application only needs to support reading secrets from files, which is a pretty standard approach.
+
+|  | Platform Effort | Deployment Effort | Application Effort |
+| --- | --- | --- | --- |
+| Vault Sidecar Injector | ■■□□□ | ■■■■■ | ■■□□□ |
+
+#### Vault Sidecar Injector: Support
+
+The Agent Sidecar Injector uses the [Vault Agent](https://www.notion.so/Comparing-methods-for-accessing-secrets-in-HashiCorp-Vault-from-Kubernetes-f05ff78d48604a9392c5643fb86bed62) under the hood which seems to support all secrets engines and all auth methods. It’s not very clear exactly what is and is not supported but the Vault Agent is quite general purpose so you would assume the support for multiple secret engines and auth methods is high.
+
+The Sidecar Injector cannot synchronise to Kubernetes secrets. This can make it quite tricky to integrate with 3rd party tools that do not make it easy to read from files, or 3rd party tools where the only choice you have is to provide a command line argument or an environment variable. People come up with all kinds of [happy hacks](https://github.com/hashicorp/vault-k8s/issues/14#issuecomment-1073361921) to work around this, but yeah, probably you should look at another approach if that’s your issue.
```

Comment:
```suggestion
The Sidecar Injector cannot synchronise to Kubernetes secrets. This can make it quite tricky to integrate with 3rd party tools that do not provide an easy method to read from files, or 3rd party tools where the only option you have is to provide either a command line argument or an environment variable. The community has invented several [happy hacks](https://github.com/hashicorp/vault-k8s/issues/14#issuecomment-1073361921) to work around this, but you should probably look for another approach if that’s an issue for you.
```